xiaomi.vacuum.gen1/firmwarebuilder/flasher.py

```diff
@@ -15,63 +44,128 @@ def md5(fname):
-    ip = m.group(1)
-except:
-    print('Vacuum not found...Check connection')
+def select_item(welcome_text, items):
+    print(welcome_text)
+    for i, item in enumerate(items):
+        print('{}. {}'.format(i+1, item))
+    try:
+        selected = input('Please select option by typing number (1-{}): '.format(len(items)))
+        result = items[int(selected)-1]
+        return result
+    except KeyboardInterrupt:
+        print('User requested to exit')
+        exit()
+    except ValueError:
+        print('Error! Please enter only one number')
+        exit()
+    except IndexError:
+        print('Error! Please enter one number between 1-{}'.format(len(items)))
+        exit()
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Flasher for Xiaomi Vacuum.\nFor specific options check \'{} --help\''.format(sys.argv[0]))
+    parser.add_argument('-a', '--address', dest='address', type=str, help='IP address of vacuum cleaner')
+    parser.add_argument('-t', '--token', dest='token', type=str, help='Known token of vacuum')
+    parser.add_argument('-f', '--firmware', dest='firmware', type=str, help='Path to firmware file')
+    parser.add_argument('-p', '--port', dest='http_port', type=str, default='8080', help='Port for local http server')
+
+    args, external = parser.parse_known_args()
+
+    print('Flasher for Xiaomi Vacuum')
+    print('Searching for device...')
+
+    ip_address = args.address
+    known_token = args.token
+    firmware = args.firmware
+
+    if not args.firmware:
+        print('You should specify firmware file name to install')
+        exit()
+
+    if not ip_address:
+        print('Address is not set. Trying to discover.')
+        timeout = 5
+        seen_addrs = []  # type: List[str]
+        addr = '<broadcast>'
+        # magic, length 32
+        helobytes = bytes.fromhex('21310020ffffffffffffffffffffffffffffffffffffffffffffffffffffffff')
+
+        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
+        s.settimeout(timeout)
+        s.sendto(helobytes, (addr, 54321))
+        while True:
+            try:
+                data, addr = s.recvfrom(1024)
+                if addr[0] not in seen_addrs:
+                    seen_addrs.append(addr[0])
+            except socket.timeout:
+                break  # ignore timeouts on discover
+            except Exception as ex:
+                print('Error while reading discover results:', ex)
+                break
+
+        if len(seen_addrs) == 0:
+            print('No devices discovered.')
+            exit()
+        elif len(seen_addrs) == 1:
+            ip_address = seen_addrs[0]
+        else:
+            ip_address = select_item('Choose device for connection:', seen_addrs)
+
+    vacuum = miio.Vacuum(ip=ip_address, token=known_token)
+    if not known_token:
+        m = vacuum.do_discover()
+        vacuum.token = m.checksum
+
+    try:
+        vacuum.find()
```

Comment:
this will cause the robot to give audible output of its whereabouts, if that's not wanted you could simply use status here.

Reply:
I can't decide myself if audible is necessary or not. Probably it is not. An option? Too many options for sure. TBD anyway.

Reply:
Imo it isn't worth to add an option for that, I just wanted to point it out as I wasn't sure what you wanted to have here.